lib/Backends/CPU/LLVMIRGen.cpp

```diff
@@ -672,6 +676,52 @@ static bool isOverlappingWithAnyBundleBufferOperands(
 void LLVMIRGen::generateLLVMIRForModule(llvm::IRBuilder<> &builder) {
   // Go over the instructions and try to group them into bundles.
   auto &instrs = F_->getInstrs();
+  // Number of IR instructions emitted into the current LLVM function.
+  size_t numInstrsInCurrentFn = 0;
+  // Max number of IR instructions to be emitted into a single LLVM function.
+  constexpr size_t maxNumInstrsPerFn = 500;
+  // Current LLVM function that should be used for emitting LLVM instructions.
+  llvm::Function *curFn = builder.GetInsertBlock()->getParent();
+
+  // Create a new LLVM function with the same signature as the current one.
+  // Call this function and pass the arguments of the current function down to
+  // the new one.
+  // This is done to keep all generated LLVM functions rather small
+  // to speed up the JIT compilation.
+  auto createNewFnIfNecessary = [&]() {
+    // Check if the current function is getting too big and a new function needs
+    // to be created.
+    if (numInstrsInCurrentFn > maxNumInstrsPerFn) {
+      auto int8PtrTy = llvm::Type::getInt8PtrTy(ctx_);
+      auto sizeTPtrTy = llvm::Type::getIntNPtrTy(ctx_, sizeof(size_t) * 8);
+      // Type of the return value.
+      llvm::Type *returnTy = llvm::Type::getVoidTy(ctx_);
+      // Types of parameters.
+      llvm::Type *paramTys[] = {int8PtrTy, int8PtrTy, int8PtrTy, sizeTPtrTy};
+      llvm::FunctionType *jitFuncTy =
+          llvm::FunctionType::get(returnTy, paramTys, /* isVarArg */ false);
+      auto *func = llvm::Function::Create(
+          jitFuncTy, llvm::Function::ExternalLinkage, "main", llmodule_.get());
+      // This function should not be inlined as it would lead to huge functions.
+      func->addFnAttr(llvm::Attribute::NoInline);
+      std::vector<llvm::Value *> args;
+      for (auto &arg : curFn->args()) {
+        args.emplace_back(&arg);
+      }
```

Comment:
How do we make sure we capture all live-in values here?

I guess it works because all our data dependencies go through Variables, which are globals, but if we ever change that, this doesn't work, right?

Reply:
Yes, all data dependencies go through memory buffers for activations and weights. There are no live-in values currently.